src/main/java/com/aws/greengrass/deployment/ThingGroupHelper.java

```diff
@@ -0,0 +1,88 @@
+    public static final String THING_GROUP_RESOURCE_TYPE = "thinggroup";
+    public static final String THING_GROUP_RESOURCE_TYPE_PREFIX  = THING_GROUP_RESOURCE_TYPE + "/";
+    private final GreengrassServiceClientFactory clientFactory;
+    private final DeviceConfiguration deviceConfiguration;
+
+    private final RetryUtils.RetryConfig clientExceptionRetryConfig =
+            RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofMinutes(1))
+                    .maxRetryInterval(Duration.ofMinutes(1)).maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Arrays.asList(SdkClientException.class)).build();
+
+    @Inject
+    public ThingGroupHelper(GreengrassServiceClientFactory clientFactory, DeviceConfiguration deviceConfiguration) {
+        this.clientFactory = clientFactory;
+        this.deviceConfiguration = deviceConfiguration;
+    }
+
+    /**
+     * Retrieve the thing group names the device belongs to.
+     * @return list of thing group names
+     * @throws InterruptedException if the thread is interrupted when fetching thing group list
+     * @throws NonRetryableDeploymentTaskFailureException  when not able to fetch thing group names
+     */
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.AvoidRethrowingException"})
+    public Optional<Set<String>> listThingGroupsForDevice()
+            throws InterruptedException, NonRetryableDeploymentTaskFailureException {
+
+        if (!deviceConfiguration.isDeviceConfiguredToTalkToCloud()) {
+            return Optional.empty();
+        }
+        try {
+            AtomicReference<String> nextToken = new AtomicReference<>();
+            Set<String> thingGroupNames = new HashSet<>();
+            return RetryUtils.runWithRetry(clientExceptionRetryConfig,
+                    () -> {
+                        do {
+                            ListThingGroupsForCoreDeviceRequest request = ListThingGroupsForCoreDeviceRequest.builder()
+                                    .coreDeviceThingName(Coerce.toString(deviceConfiguration.getThingName()))
+                                    .nextToken(nextToken.get()).build();
+                            ListThingGroupsForCoreDeviceResponse response = clientFactory.getGreengrassV2DataClient()
+                                    .listThingGroupsForCoreDevice(request);
+                            response.thingGroups().forEach(thingGroup -> {
+                                //adding direct thing groups
+                                thingGroupNames.add(THING_GROUP_RESOURCE_TYPE_PREFIX + thingGroup.thingGroupName());
+                                //adding parent thing groups
+                                thingGroup.rootToParentThingGroups().forEach(parentGroup -> thingGroupNames
+                                                .add(THING_GROUP_RESOURCE_TYPE_PREFIX + parentGroup.thingGroupName()));
+                            });
```

Comment:
Where did we land on handling deployments to child and parent groups with conflicting versions of components? I remember some discussion around it

Reply:
We haven't prioritized this yet.